substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfLocSectionImpl.java

```diff
@@ -0,0 +1,718 @@
+            if (buffer == null) {
+                pos += putShort(byteCount, scratch, 0);
+                pos += putByte(regOp, scratch, 0);
+                pos += putULEB(targetIdx, scratch, 0);
+            } else {
+                pos = putShort(byteCount, buffer, pos);
+                pos = putByte(regOp, buffer, pos);
+                pos = putULEB(targetIdx, buffer, pos);
+                verboseLog(context, "  [0x%08x]     REGOP count %d op 0x%x reg %d", pos, byteCount, regOp, targetIdx);
+            }
+            // target idx written as ULEB should fit in one byte
+            assert pos == p + 4 : "wrote the wrong number of bytes!";
+        }
+        return pos;
+    }
+
+    private int writeStackLocation(DebugContext context, int offset, byte[] buffer, int p) {
+        int pos = p;
+        short byteCount = 0;
+        int sp = getDwarfStackRegister();
+        byte stackOp;
+        if (sp < 32) {
+            // fold the base reg index into the op
+            stackOp = DwarfDebugInfo.DW_OP_breg0;
+            stackOp += sp;
+        } else {
+            // pass base reg index as a ULEB operand
+            stackOp = DwarfDebugInfo.DW_OP_bregx;
+        }
+        if (buffer == null) {
+            pos += putShort(byteCount, scratch, 0);
+            pos += putByte(stackOp, scratch, 0);
+            if (stackOp == DwarfDebugInfo.DW_OP_bregx) {
+                // need to pass base reg index as a ULEB operand
+                pos += putULEB(sp, scratch, 0);
+            }
+            pos += putSLEB(offset, scratch, 0);
+        } else {
+            int patchPos = pos;
+            pos = putShort(byteCount, buffer, pos);
+            int zeroPos = pos;
+            pos = putByte(stackOp, buffer, pos);
+            if (stackOp == DwarfDebugInfo.DW_OP_bregx) {
+                // need to pass base reg index as a ULEB operand
+                pos = putULEB(sp, buffer, pos);
+            }
+            pos = putSLEB(offset, buffer, pos);
+            // now backpatch the byte count
+            byteCount = (byte) (pos - zeroPos);
+            putShort(byteCount, buffer, patchPos);
+            if (stackOp == DwarfDebugInfo.DW_OP_bregx) {
+                verboseLog(context, "  [0x%08x]     STACKOP count %d op 0x%x offset %d", pos, byteCount, stackOp, 0 - offset);
+            } else {
+                verboseLog(context, "  [0x%08x]     STACKOP count %d op 0x%x reg %d offset %d", pos, byteCount, stackOp, sp, 0 - offset);
+            }
+        }
+        return pos;
+    }
+
+    private int writeConstantLocation(DebugContext context, JavaConstant constant, byte[] buffer, int p) {
+        int pos = p;
+        byte op = DW_OP_implicit_value;
+        JavaKind kind = constant.getJavaKind();
+        int byteCount = (kind == JavaKind.Object ? 8 : kind.getByteCount());
+        assert (kind != JavaKind.Object || constant.isNull()) : "only expecting null object constant!";
+        if (buffer == null) {
```

Comment:
This needs fixing. The DWARF expression defining the constant needs to be preceded with a total byte count for the expression bytes. I accidentally deleted that before pushing the update for CONSTANT support.

Reply:
I corrected the error and also added support for writing object constants. So, you can now see constant strings in the initial heap etc.